pkg/operator/builder.go

```diff
@@ -0,0 +1,349 @@
+	b.namespaces = namespaces
+	return b
+}
+
+func (b *Builder) WithLogger(logger *zap.Logger) *Builder {
+	b.logger = logger
+	return b
+}
+
+func (b *Builder) WithSyncPeriod(period time.Duration) *Builder {
+	b.syncPeriod = period
+	return b
+}
+
+func (b *Builder) WithMetricAddress(address string) *Builder {
+	b.metricAddress = address
+	return b
+}
+
+func (b *Builder) WithProbeAddress(address string) *Builder {
+	b.probeAddress = address
+	return b
+}
+
+func (b *Builder) WithLeaderElection(enable bool) *Builder {
+	b.leaderElection = enable
+	return b
+}
+
+func (b *Builder) WithAtlasDomain(domain string) *Builder {
+	b.atlasDomain = domain
+	return b
+}
+
+func (b *Builder) WithPredicates(predicates []predicate.Predicate) *Builder {
+	b.predicates = predicates
+	return b
+}
+
+func (b *Builder) WithAPISecret(apiSecret client.ObjectKey) *Builder {
+	b.apiSecret = apiSecret
+	return b
+}
+
+func (b *Builder) WithAtlasProvider(provider atlas.Provider) *Builder {
+	b.atlasProvider = provider
+	return b
+}
+
+func (b *Builder) WithFeatureFlags(featureFlags *featureflags.FeatureFlags) *Builder {
+	b.featureFlags = featureFlags
+	return b
+}
+
+func (b *Builder) WithDeletionProtection(deletionProtection bool) *Builder {
+	b.deletionProtection = deletionProtection
+	return b
+}
+
+func (b *Builder) Build(ctx context.Context) (manager.Manager, error) {
+	mergeDefaults(b)
+
+	utilruntime.Must(clientgoscheme.AddToScheme(b.scheme))
+	utilruntime.Must(akov2.AddToScheme(b.scheme))
+
+	cacheOpts := cache.Options{
+		SyncPeriod: &b.syncPeriod,
+	}
+
+	switch len(b.namespaces) {
+	case 0:
+		cacheOpts.ByObject = map[client.Object]cache.ByObject{
+			&corev1.Secret{}: {
+				Label: labels.SelectorFromSet(labels.Set{
+					connectionsecret.TypeLabelKey: connectionsecret.CredLabelVal,
+				}),
+			},
+		}
+	default:
+		cacheOpts.DefaultNamespaces = map[string]cache.Config{}
+		for _, namespace := range b.namespaces {
+			cacheOpts.DefaultNamespaces[namespace] = cache.Config{}
+		}
+	}
+
+	mgr, err := ctrl.NewManager(
+		b.config,
+		ctrl.Options{
+			Scheme:  b.scheme,
+			Metrics: metricsserver.Options{BindAddress: b.metricAddress},
+			WebhookServer: webhook.NewServer(webhook.Options{
+				Port: 9443,
+			}),
+			Cache:                  cacheOpts,
+			HealthProbeBindAddress: b.probeAddress,
+			LeaderElection:         b.leaderElection,
+			LeaderElectionID:       b.leaderElectionID,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = mgr.AddHealthzCheck("health", healthz.Ping); err != nil {
+		return nil, err
+	}
+
+	if err = mgr.AddReadyzCheck("check", healthz.Ping); err != nil {
+		return nil, err
+	}
+
+	if err = indexer.RegisterAll(ctx, mgr, b.logger); err != nil {
+		return nil, fmt.Errorf("unable to create indexers: %w", err)
+	}
+
+	if b.atlasProvider == nil {
+		b.atlasProvider = atlas.NewProductionProvider(b.atlasDomain, b.apiSecret, mgr.GetClient())
+	}
+
+	if err = (&atlasproject.AtlasProjectReconciler{
```

Comment:
since we're here. one thing that I recognized for quite some time is that none of the reconciler structs has `New...` constructor methods. Please add `NewAtlasProjectReconciler(...) *atlasproject.AtlasProjectReconciler` for this and the reconcilers below.